pkgs/applications/virtualization/xen/patches.nix

```
@@ -0,0 +1,63 @@
+# Patching Xen? Check the XSAs at
+# https://xenbits.xen.org/xsa/
+# and try applying all the ones
+# we haven't gotten around to
+# yet, if any are necessary.
+
+{ fetchpatch }:
+
+let
+  xsaPatch =
+    { name, hash }:
+    (fetchpatch {
+      url = "https://xenbits.xen.org/xsa/xsa${name}.patch";
+      inherit hash;
+    });
+  qubesPatch =
+    {
+      name,
+      tag,
+      hash,
+    }:
+    (fetchpatch {
+      url = "https://raw.githubusercontent.com/QubesOS/qubes-vmm-xen/v${tag}/${name}.patch";
+      inherit hash;
+    });
+in
+{
+  # Example patches:
+  #
+  # "XSA_100" = xsaPatch {
+  #   name = "100";
+  #   hash = "sha256-0000000000000000000000000000000000000000000000000000";
+  # };
+  #
+  # "QUBES_libxl-fix-all-issues" = qubesPatch {
+  #   name = "1000-libxl-fix-all-issues";
+  #   tag = "4.20.0-1";
+  #   hash = "sha256-0000000000000000000000000000000000000000000000000000";
+  # };
+
+  # Build reproducibility patches for Xen 4.17.4, but works on 4.18.2.
+  QUBES_Define-build-dates-time-based-on-SOURCE_DATE_EPOCH = qubesPatch {
+    name = "1100-Define-build-dates-time-based-on-SOURCE_DATE_EPOCH";
+    tag = "4.17.4-3";
+    hash = "sha256-OwKA9oPTwhRcSmiOb+PxzifbO/IG8IHWlvddFh/nP6s=";
+  };
+  QUBES_docs-rename-DATE-to-PANDOC_REL_DATE-and-allow-to-spe = qubesPatch {
+    name = "1101-docs-rename-DATE-to-PANDOC_REL_DATE-and-allow-to-spe";
+    tag = "4.17.4-3";
+    hash = "sha256-BUtYt0mM3bURVaGv4oDznzxx1Wo4sfOpGV5GB8qc5Ns=";
+  };
+  QUBES_docs-xen-headers-use-alphabetical-sorting-for-incont = qubesPatch {
+    name = "1102-docs-xen-headers-use-alphabetical-sorting-for-incont";
+    tag = "4.17.4-3";
+    hash = "sha256-mQUp2w9lUb7KDq5MuPQjs6y7iuMDeXoZjDjlXfa5z44=";
+  };
+
+  # Double unlock in x86 guest IRQ handling. (4.16 - 4.19-rc3)
+  "XSA_458" = xsaPatch {
+    name = "458";
+    hash = "sha256-yHI9Sp/7Ed40iIYQ/HOOIULlfzAzL0c0MGqdF+GR+AQ=";
+  };
+}
```

Provide more understandable comment for why is this needed

---

Uhh, to patch for XSA-458? Seems pretty self-explanatory. The comment refers to the XSA name, and the versions that may receive this patch.

---

```suggestion
  # Double unlock in x86 guest IRQ handling for "Multiple Messages" feature of PCI MSI by not passing multi-vector MSI capable devices to x86 guests (4.16 - 4.19-rc3)
  "XSA_458" = xsaPatch {
    name = "458";
    hash = "sha256-yHI9Sp/7Ed40iIYQ/HOOIULlfzAzL0c0MGqdF+GR+AQ=";
  };
}
```

I would prefer this to make it immediately obvious why it's there and what it's doing instead of the generic title of the fix.

---

I don't even think the basic title as a comment is necessary. Just the XSA number should tell you all the information you need to figure out what is the vulnerability being patched here. This one-line comment will never be as descriptive as the full advisory.

---

@SigmaSquadron Best to expand "Xen Security Advisory" in a comment ahead of randomly using XSA as a name.  That makes it clear that it's a security fix, and from there it's not hard to work out it's a security fix beyond the base tag you import.

---

Here's what I propose, which also considers @LeSuisse's request above:
```suggestion
  # Xen Security Advisory #458: (4.16 - 4.19-rc3)
  #
  #                Double unlock in x86 guest IRQ handling
  # An optional feature of PCI MSI called "Multiple Message" allows a device
  # to use multiple consecutive interrupt vectors.  Unlike for MSI-X, the
  # setting up of these consecutive vectors needs to happen all in one go.
  # In this handling an error path could be taken in different situations,
  # with or without a particular lock held. This error path wrongly releases
  # the lock even when it is not currently held.
  "XSA_458" = xsaPatch {
    id = "458";
    name = "Double unlock in x86 guest IRQ handling";
    cve = "CVE-2024-31143";
    hash = "sha256-yHI9Sp/7Ed40iIYQ/HOOIULlfzAzL0c0MGqdF+GR+AQ=";
  };
```

---

I would also include a URL to the security advisory to make it just click away instead of having to go through the search engine

---

Added the `meta.cveUrl` attribute to the patch. It's populated automatically from the `cve` attribute.

---

So that means that i have to go through `nix repl` to get that URL, if i want to avoid using the search engine? If so it doesn't feel like a good solution from a code review perspective.

---

I'm sorry, but I'm not writing anything more complex than that in order to save people two keyboard shortcuts into the browser.

---

So, it turns out you cannot add arbitrary meta attributes. Unfortunate.

---

> I'm sorry, but I'm not writing anything more complex than that in order to save people two keyboard shortcuts into the browser.

So, I lied, and ended up over-engineering something very nice! The meta.longDescription now includes information about which XSAs this Xen is protected against. This addition is dynamic and will automatically change depending on which version is being built, and what XSA patches are being applied.

One caveat with this change is that multiple patches for one XSA will no longer be supported. If that happens, I'll just comment-out these changes and go back to the original behaviour, as the actual security of this package obviously has priority over fancy descriptions. Until then, we can hope Xen doesn't publish XSAs with multiple patch files, or someone more knowledgeable than me can come in and fix the mess I've made.

---

It looks better, but that includes that from a user perspective, proposed https://github.com/NixOS/nixpkgs/pull/324693#discussion_r1686233973